### PR TITLE
[#2759] Delay server now honors a rule's priority (master)

### PR DESCRIPTION
**Pull request by korydraughn**
Valid priority levels are defined by the following range: [1 - 9]

The delay server gathers and schedules rules in descending order of
priority level.

`<PRI>` tag is no longer supported by the delay(). Use `<PRIORITY>` instead.
Attempting to use `<PRI>` will result in an error.

Reoccurring rules that have an empty priority level will be updated to
have a priority level of 5 following successful execution. This also
applies to rules that have an invalid priority level.

### CODE REVIEW

**Review by trel — COMMENTED** [review]
and how do we know this makes the delay server run the rules in this order?    is that already handled/known?

**Comment by korydraughn** [comment]
The rules are **scheduled** in the correct order. They aren't guaranteed to finish in that order due to the OS.

File: server/api/src/rsRuleExecSubmit.cpp [hunk 1]
```diff
@@ -52,6 +55,24 @@ namespace
         irods::experimental::key_value_proxy kvp{ruleExecSubmitInp->condInput};
         kvp[RULE_EXECUTION_CONTEXT_KW] = irods::to_json(rei_info->rei).dump().data();
 
+        // Verify that the priority is valid if the client provided one.
+        if (std::strlen(ruleExecSubmitInp->priority) > 0) {
+            try {
+                if (const auto p = std::stoi(ruleExecSubmitInp->priority); p < 1 || p > 9) {
+                    rodsLog(LOG_ERROR, "Delay rule priority must satify the following requirement: 1 <= P <= 9.");
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply server/api/src/rsRuleExecSubmit.cpp]: done

**Comment by trel** [comment]
code looks good.   do we need a test to prove the ordering is honored?

**Comment by korydraughn** [comment]
Yes. I am planning to write all of those next. I hope to have them done some time tomorrow.

**Comment by korydraughn** [comment]
The tests will focus on verifying that the priority is set/updated in the correct situations (e.g. set the priority to 5 if it is empty). It will be difficult to impossible to verify execution order, so there won't be tests for that.

**Comment by korydraughn** [comment]
added #

**Pull request merged**